### ISSUE TITLE
language files

### ISSUE THREAD

**Issue by KarelWintersky** [PHPAuth/PHPAuth#89]
I think about loading language file while config initialization.. and name of language file we can take from config value like `site_messages_language`.

Of course, if is_file() returns false, we will load `en_GB.php`

Also, I think about storing different translations in JSON-file, but.. I can be not a good idea.

**Comment by Conver** [comment]
That would indeed be a good idea.

Why are you thinking about JSON based files?

**Comment by KarelWintersky** [comment]
For now I'm weighting the options on the implementation. Both approaches have their ups and downs, I'm not ready to give concrete answers.

**Comment by Conver** [comment]
Lets add this to the todo, maybe other can come with suggestions too

**Comment by KarelWintersky** [comment]
don't close this issue, ok?

**Comment by Conver** [comment]
I won't :)

**Comment by KarelWintersky** [comment]
```php
$this->config['site_language'] 
= (is_null($this->config['site_language'])) 
? 'en_GB' 
: $this->config['site_language'];

require_once("languages/{$this->config["site_language"]}.php");
$this->lang = $lang;
```

This construction should not work:
`Notice: Undefined variable: lang in /phpauth/config.class.php on line 40`

Also, 
```
Notice: Undefined index: lang in /phpauth/config.class.php on line 59
```

**Comment by KarelWintersky** [comment]
Out code must be:

```php
public function __construct(\PDO $dbh, $config_table = 'config')
    {
        $this->dbh = $dbh;

        if (func_num_args() > 1)
            $this->phpauth_config_table = $config_table;

        $this->config = array();

        $query = $this->dbh->query("SELECT * FROM {$this->config_table}");

        while($row = $query->fetch()) {
            $this->config[$row['setting']] = $row['value'];
        }

        $this->setVerifyDefaults(); // Danger foreseen is half avoided.

        $this->config['site_language']
            = (isset($this->config['site_language']) && is_null($this->config['site_language']))
            ? 'en_GB'
            : $this->config['site_language'];

        $lang_file = dirname(__FILE__).DIRECTORY_SEPARATOR."languages/{$this->config["site_language"]}.php";

        require_once($lang_file);

        $this->config['lang'] = $lang;
    }

```

and `auth.php`:
```php

public function __construct(\PDO $dbh, $config, $lang)
	{
		$this->dbh = $dbh;
		$this->config = $config;
		$this->lang = $this->config->lang;

        if (version_compare(phpversion(), '5.4.0', '<')) {
            die('PHP 5.4.0 required for PHPAuth engine!');
        }

		if (version_compare(phpversion(), '5.5.0', '<')) {
			require("files/password.php");
		}
	}
```

**Comment by KarelWintersky** [comment]
Also, we need implementate option - enable or disable ban for too frequent login attempts.

**Comment by Conver** [comment]
Of course you are getting it wrong when you are using the wrong class, I've added it in auth class, not config class.. the language ain't used in there.. 

Your's example is a bit overkill, thats correct.. But my example does work.. running with it here..

but anyhow.. We need to come up with a good option.

"enable or disable ban for too frequent login attempts." -> good idea..

**Comment by KarelWintersky** [comment]
yah, my inadvertence.
of course, auth.php: 
```php
public function __construct(\PDO $dbh, $config)
...
```

**Comment by mooztik** [comment]
i m using HTTP_ACCEPT_LANGUAGE to define defaut language, but if the user have defined his own, the detected one is ignored. 
PhpAuth support multi language, but it's not its role to control the language used.

**Comment by JacquesLoubser** [comment]
"enable or disable ban for too frequent login attempts." -> disabling it would open your system to brute force.(which is not a good idea) #99 helps to tell bots apart from forgetful users.

I agree with @mooztik The user of the library might set the language with a cookie, db, HTTP_ACCEPT_LANGUAGE, Geo-Location (IP or GPS) or other options.

**Comment by KarelWintersky** [comment]
> "enable or disable ban for too frequent login attempts." -> disabling it would open your system to brute force.(which is not a good idea)

I know. 
I mean, it can be **option** : ON/OFF.

**Comment by Conver** [comment]
How is it going with this one?

**Comment by cuonic** [comment]
Language shouldn't be something handled by the class, the idea is that you pass whatever language file the user wants to the PHPAuth class, so all language file choice is handled outside of the class, because language choice is installation dependant: some users only want one language available, some store the language in a cookie, some use the `HTTP_ACCEPT_LANGUAGE` header... Too many use cases for PHPAuth to handle.

**Comment by mooztik** [comment]
@cuonic i agree and i think a language handler can be set in the plugin project :D 
then the language files can finally be any format

**Comment by livmackintosh** [comment]
@mooztik I think that makes the most sense :) Given that most people won't use a persistant database connection, the Config and Auth classes will be instantiated on each page load. So, instead of passing the $lang array to the Config class, you could pass the locale code as a string (e.g "en_GB") and have a method to load the correct file - which of course has the same name as the string passed (without the extention) :smile_cat: .

Liv

**Comment by mooztik** [comment]
@livmackintosh i don't talk about using config class and config data table to handle language but to let PHPAuth use an internal plugin to handle it's own language files.
free to any developer to use the default one (php array) or implement any other format (like gettext, yml, json...) within plugin extension.
like i said, i'm currently using HTTP_ACCEPT_LANGUAGE and set the language used from that with a little function. (who load language file).
i don't need phpauth to handle this.
but if this can be handled by a plugin, i keep my free choise on how to implement language support.

**Comment by KarelWintersky** [comment]
Close issue. Found better solution.
